pkg/controller/endpointslice/endpointslice_controller_test.go

```diff
@@ -327,77 +330,494 @@ func TestOnEndpointSliceUpdate(t *testing.T) {
+						PodIPs: []v1.PodIP{
+							{
+								IP: "10.0.0.2",
+							},
+						},
+						Conditions: []v1.PodCondition{
+							{
+								Type:   v1.PodReady,
+								Status: v1.ConditionTrue,
+							},
+						},
+					},
+				},
+			},
+			expectedEndpointPorts: []discovery.EndpointPort{
+				{
+					Name:     utilpointer.StringPtr("sctp-example"),
+					Protocol: protoPtr(v1.ProtocolSCTP),
+					Port:     utilpointer.Int32Ptr(int32(3456)),
+				},
+				{
+					Name:     utilpointer.StringPtr("udp-example"),
+					Protocol: protoPtr(v1.ProtocolUDP),
+					Port:     utilpointer.Int32Ptr(int32(161)),
+				},
+				{
+					Name:     utilpointer.StringPtr("tcp-example"),
+					Protocol: protoPtr(v1.ProtocolTCP),
+					Port:     utilpointer.Int32Ptr(int32(80)),
+				},
+			},
+			expectedEndpoints: []discovery.Endpoint{
+				{
+					Conditions: discovery.EndpointConditions{
+						Ready:       utilpointer.BoolPtr(true),
+						Terminating: utilpointer.BoolPtr(false),
+					},
+					Addresses: []string{"10.0.0.1"},
+					TargetRef: &v1.ObjectReference{Kind: "Pod", Namespace: "default", Name: "pod0"},
+					Topology:  map[string]string{"kubernetes.io/hostname": "node-1"},
+				},
+				{
+					Conditions: discovery.EndpointConditions{
+						Ready:       utilpointer.BoolPtr(true),
+						Terminating: utilpointer.BoolPtr(true),
+					},
+					Addresses: []string{"10.0.0.2"},
+					TargetRef: &v1.ObjectReference{Kind: "Pod", Namespace: "default", Name: "pod1"},
+					Topology:  map[string]string{"kubernetes.io/hostname": "node-1"},
+				},
+			},
+			terminatingGateEnabled: true,
+		},
+		{
+			name: "Terminating pods with EndpointSliceTerminatingCondition disabled",
+			service: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              "ipv4",
+					Namespace:         "default",
+					CreationTimestamp: creationTimestamp,
+				},
+				Spec: v1.ServiceSpec{
+					Ports: []v1.ServicePort{
+						{Name: "tcp-example", TargetPort: intstr.FromInt(80), Protocol: v1.ProtocolTCP},
+						{Name: "udp-example", TargetPort: intstr.FromInt(161), Protocol: v1.ProtocolUDP},
+						{Name: "sctp-example", TargetPort: intstr.FromInt(3456), Protocol: v1.ProtocolSCTP},
+					},
+					Selector: map[string]string{"foo": "bar"},
+					IPFamily: &ipv4Family,
+				},
+			},
+			pods: []*v1.Pod{
+				{
+					TypeMeta: metav1.TypeMeta{APIVersion: "v1"},
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace:         "default",
+						Name:              "pod0",
+						Labels:            map[string]string{"foo": "bar"},
+						DeletionTimestamp: nil,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{{
+							Name: "container-1",
+						}},
+						NodeName: "node-1",
+					},
+					Status: v1.PodStatus{
+						PodIP: "10.0.0.1",
+						PodIPs: []v1.PodIP{{
+							IP: "10.0.0.1",
+						}},
+						Conditions: []v1.PodCondition{
+							{
+								Type:   v1.PodReady,
+								Status: v1.ConditionTrue,
+							},
+						},
+					},
```

Comment:
I think it's easier to make sense of what's changing in tests like this if things that are consistent are mostly hidden. As an example, maybe you could use a `standardPodSpec` var since that seems to be irrelevant to all of these tests. Similarly, maybe it would help to add some kind of helper func for generating `ObjectMeta`. Maybe something like this:

```suggestion
					ObjectMeta: standardPodMeta("pod0", nil),
					Spec: standardPodSpec,
					Status: v1.PodStatus{
						PodIP: "10.0.0.1",
						PodIPs: []v1.PodIP{{
							IP: "10.0.0.1",
						}},
						Conditions: []v1.PodCondition{
							{
								Type:   v1.PodReady,
								Status: v1.ConditionTrue,
							},
						},
					},
```

Reply:
I was thinking we would add tests here in the future where the labels could vary between the pods which was why I didn't try to hide any of the specific fields. I think it's safe to remove TypeMeta though.

Reply:
For standardPodSpec, I don't think any of the unit tests here use named pod ports. If we added those the pod spec would change as well :thinking:

Reply:
👍 makes sense